### ISSUE TITLE
Higher-rank types are weirdly not instantiated

### ISSUE THREAD

**Issue by Nadrieril** [gluon-lang/gluon#603]
The following code:
```
type Proxy a = | Proxy
let foo: forall b. (forall a. Proxy a) -> Proxy b
    = \p -> p
```
gives the following two errors:
```
error: Expected the following types to be equal
Expected: rankn.Proxy b
Found: forall a . rankn.Proxy a
1 errors were found during unification:
Could not generalize the variable bound to `b` as the variable was used outside its scope
- <rankn>:4:13
4 |     = \p -> p
  |             ^

error: Expected the following types to be equal
Expected: (forall a . rankn.Proxy a) -> rankn.Proxy b
Found: forall c . (forall a . rankn.Proxy a) -> c
1 errors were found during unification:
Types do not match:
    Expected: (forall a . rankn.Proxy a) -> rankn.Proxy b
    Found: forall c . (forall a . rankn.Proxy a) -> c
- <rankn>:4:7
4 |     = \p -> p
  |       ^^^^^^^
```
Instead of instanciating the forall-quantified type, it seems to be trying to generalize the bound type variable. I have no clue what the second error is on about.
Moveover, there is no way to help the compiler, because adding a type annotation fails similarly.
A related issue is that annotating a bound polymorphic variable with its own type fails too:
```
let foo id: (forall a. a -> a) -> () = 
    let id': forall a. a -> a = id
    in ()
```

I don't know if this is a know case that the type system cannot handle or a bug, but I found it surprising (and annoying).

**Comment by Marwes** [comment]
I fixed the other two issues with higher-rank issues. This one will need some extra work to make sure that I don't introduce any unsoundness. For instance, it interacts closely with the following where I don't want the type `forall m . m` to be propagated to the argument `m` of `make`.

```f#
let make m =
    let m2: forall m . m = m
    let _ = m2 1
    m
make 2
```
https://github.com/gluon-lang/gluon/blob/2562070dd81499cff7ff076e8bb5e46c7a6692a4/check/tests/fail.rs#L283-L288

**Comment by Nadrieril** [comment]
Thanks a lot !

**Comment by Marwes** [comment]
Should have been closed by 67a502f   🤷‍♂️
